router_client.go

```diff
@@ -560,3 +646,203 @@ func (r *routerClient) SubscribeHtlcEvents(ctx context.Context) (
 
 	return htlcChan, errChan, nil
 }
+
+// InterceptHtlcs intercepts htlcs on a node, using the handler function
+// provided to provide the interceptor with interception decisions. The handler
+// provided may block, but must exit if the context passed in is canceled. It
+// is also responsible for releasing htlcs on exit.
+//
+// There are a few ways in which this method can exit:
+// - ctx canceled: the calling client cancels
+// - r.quit: the router is shut down
+// - lnd stream error: lnd has exited
+// - handler error: a critical error occurred while handing a htlc
+func (r *routerClient) InterceptHtlcs(ctx context.Context,
+	handler HtlcInterceptHandler) error {
+
+	// Create a child context that will be canceled when this function
+	// exits. We use this context to be able to cancel goroutines when we
+	// exit on errors, because the parent context won't be canceled in that
+	// case.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	stream, err := r.client.HtlcInterceptor(
+		r.routerKitMac.WithMacaroonAuth(ctx),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Create an error channel that we'll send errors on if any of our
+	// goroutines fail. We buffer by 1 so that the goroutine doesn't depend
+	// on the stream being read, and select on context cancelation and
+	// quit channel so that we do not block in the case where we exit with
+	// multiple errors.
+	errChan := make(chan error, 1)
+
+	sendErr := func(err error) {
+		select {
+		case errChan <- err:
+		case <-ctx.Done():
+		case <-r.quit:
+		}
+	}
+
+	// Start a goroutine that consumes interception requests from lnd and
+	// sends them into our requests channel for handling. The requests
+	// channel is not buffered because we expect all requests to be handled
+	// until this function exits, at which point we expect our context to
+	// be canceled or quit channel to be closed.
+	requestChan := make(chan InterceptedHtlc)
+	r.wg.Add(1)
```

Comment:
There's a (very small) chanche that someone calls `InterceptHtlcs` while the `LndServices` cleanup function is called. In this case there's an even smaller chance that we intercept an htlc but we quit right after that (selecting on the `quit` channel) never sending the htlc to the `handler`.  A possible fix could be to always send intercepted htlcs to the `requestChan` and then before shutting down any non handled htlcs should be released inside the other for loop (or maybe in the `stop` function). wdyt?

Reply:
I was a bit unsure of the internals of the interceptor when I wrote this (hence all the scattered todos). 

Spent some time looking into it, and it seems that the [interceptor](https://github.com/lightningnetwork/lnd/blob/master/lnrpc/routerrpc/forward_interceptor.go#L69) will always [release](https://github.com/lightningnetwork/lnd/blob/master/lnrpc/routerrpc/forward_interceptor.go#L210) all our currently held htlcs, so I think it's ok if we don't pass htlcs on to the handler in our shutdown case? 

Will update the comments in this PR to reflect this, requested review before coming back and updating w/ my new understanding of the interceptor, sorry!

Reply:
Indeed that's reassuring! Thanks for pointing to the relevant part.